rfc-123-github-actions-ci.md

```diff
@@ -50,6 +50,20 @@ This would be a gradual process, because we have many repos, but this decision w
 - Another system to learn, as well as the replatforming work already ongoing.
 - Duplication of build parameters and steps between all our apps, as they’re all Rails apps.
 
+### Findings from some initial explorations into using GitHub Actions
+
+#### `push` vs `pull_request` build triggers
+
+There has been some [chatter](https://github.com/alphagov/govuk-developer-docs/pull/2497#discussion_r415572147) about if we want `push` or `pull_request` triggers for tests to run, or both. The GOV.UK Developer Tools group has done some investigation into the differences between this, as there was some concern about if we wanted long-running builds (in the case of apps with large test suites, like whitehall) to run twice when not needed. There is not currently any way to differentiate and say "run on PRs from forks" and "run on push events to main repo branches". We'd have really liked a `pull_request_only_from_forks` setting, but that doesn't exist. It's worth noting that this isn't a problem _only_ with GitHub Actions: [Travis](https://github.com/travis-ci/travis-ci/issues/1147) has had the same pain points for years, and we currently do double builds for the apps we use Travis for.
+
+Some solutions to this include:
+
+- Running `on: [pull_request, push]`. This means that internal devs get feedback on their commits when they push them, for example if they've forgotten to run RuboCop locally. And also so that external contributors PRs get CI feedback on their changes. It does mean that for long-running test suites, we could waste compute power, time and the planet's precious natural resources running things twice that don't need to be run twice.
```

Comment:
I've come around to be in favour of this option since they're slightly different changes. Does seem a pain that we might often have very similar tests run - but at least they'd be in parallel. 

I feel like this would become much more of a pain point though if action queues get long and we find we're waiting for one and not the other.

Reply:
On the Enterprise plan, we get [180 concurrent builds](https://help.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits) across `alphagov`, assuming they're all Linux builds and not macOS.

Reply:
At the end of the day, I don't mind which option we go for (as long as it's one of "only PRs" or "both PRs and pushes", the "on push" option is in no way an improvement to our current processes), but I'd like to solicit some more opinions from others otherwise this feels a bit circular. :-)

Reply:
I can see value in having test run on your branch before a PR is raised, therefore you can immediately notice a failure, rather than waiting after you've raised the PR. I feel like going with both lets us see if there's any value/downsides and we can always tweak these settings in the future.

Reply:
It's worth describing that advantage in the RFC 👍

Reply:
Have we considered another potential disadvantage: if we're running the tests twice, and they (both) fail, which one do I look at to see the failure? While this may be obvious for us, the architects of the future, I can see it being a source of confusion: which "Details" link should I click on when I see a failure?

Reply:
If the failures are different then we have more interesting problems. I'd look at the `push` ones, personally, as I'll be able to see quicker that it's failed. Unless it's a forked PR.